### ISSUE TITLE
Spurious EIO with zfs-2.1.8 (bisected to 0156253d29a303bdcca3e535958e754d8f086e33)

### ISSUE THREAD

**Issue by thesamesam** [openzfs/zfs#14413]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| Gentoo
Distribution Version	| Rolling
Kernel Version	| 5.15.89-gentoo-dist-hardened
Architecture	| amd64
OpenZFS Version	| 2.1.8 (2.1.7 works)

### Describe the problem you're observing

On an encrypted dataset (details below), I can't seem to clone a specific git repo:
```
$ zfs get encryption zroot
NAME   PROPERTY    VALUE        SOURCE
zroot  encryption  aes-256-gcm  -
$ zfs get encryption zroot/HOME
NAME        PROPERTY    VALUE        SOURCE
zroot/HOME  encryption  aes-256-gcm  -
$ zfs get encryption zroot/HOME/sam
NAME            PROPERTY    VALUE        SOURCE
zroot/HOME/sam  encryption  aes-256-gcm  -
```
```
$ git clone https://github.com/Warzone2100/warzone2100
[...]
remote: Total 187467 (delta 15), reused 14 (delta 9), pack-reused 187437
Receiving objects: 100% (187467/187467), 665.19 MiB | 15.16 MiB/s, done.
warning: die() called many times. Recursion error or racy threaded death!
fatal: cannot pread pack file: Input/output error
atal: cannot pread pack file: Input/output error
fatal: fetch-pack: invalid index-pack output
```

I've bisected to 0156253d29a303bdcca3e535958e754d8f086e33 and reverting that seems to work.

### Describe how to reproduce the problem

1. Build ZFS 2.1.8
2. Try `git clone https://github.com/Warzone2100/warzone2100` on an encrypted dataset?

### Include any warning/errors/backtraces from the system logs

None. Let me know how I can obtain any useful ones.

**Comment by thesamesam** [comment]
cc @ryao (backport PR, https://github.com/openzfs/zfs/pull/14366)
cc @ahrens (original commit, 40d7e971ffc16b2ef993a6e9da40a8b3ca91ad01)

**Comment by gyakovlev** [comment]
@thesamesam asked me to test it a bit as well, I can't reproduce unfortunately.
I tried to clone to encrypted dataset (top level)
also tried to clone to a following config:

toplevel(encrypted)/leafdataset(different key from top level)
toplevel(encrypted)/subdataset(unencrypted)/leafdataset(different key from top level)

I'm on ppc64le tho, idk if it matters. maybe it does not take some codepath in my case?

**Comment by rincebrain** [comment]
Hey @tonyhutter you might want to cut a 2.1.8.1 with just this reverted because getting EIO back when there's no IO problem seems Very Bad.

**Comment by tonyhutter** [comment]
I put out a small fix-up zfs-2.1.9 release PR with 0156253d29a303bdcca3e535958e754d8f086e33 reverted:

https://github.com/openzfs/zfs/pull/14427

**Comment by tonyhutter** [comment]
[zfs-2.1.9](https://github.com/openzfs/zfs/releases/tag/zfs-2.1.9) is released with https://github.com/openzfs/zfs/commit/0156253d29a303bdcca3e535958e754d8f086e33 reverted.

**Comment by ahrens** [comment]
@thesamesam and anyone else who is experiencing this, can you provide the output of `zfs get all <dataset>`?

**Comment by thesamesam** [comment]
I can't reproduce this right now with the original commit applied (on top of revert), but my affected pool is definitely quite fragmented.

```
$ zfs get all zroot
NAME   PROPERTY              VALUE                           SOURCE
zroot  type                  filesystem                      -
zroot  creation              Thu Aug 20 20:05 2020           -
zroot  used                  366G                            -
zroot  available             67.6G                           -
zroot  referenced            98K                             -
zroot  compressratio         1.50x                           -
zroot  mounted               no                              -
zroot  quota                 none                            default
zroot  reservation           none                            default
zroot  recordsize            128K                            default
zroot  mountpoint            none                            local
zroot  sharenfs              off                             default
zroot  checksum              on                              default
zroot  compression           zstd                            local
zroot  atime                 off                             local
zroot  devices               on                              default
zroot  exec                  on                              default
zroot  setuid                on                              default
zroot  readonly              off                             default
zroot  zoned                 off                             default
zroot  snapdir               hidden                          default
zroot  aclmode               discard                         default
zroot  aclinherit            restricted                      default
zroot  createtxg             1                               -
zroot  canmount              on                              local
zroot  xattr                 sa                              local
zroot  copies                1                               default
zroot  version               5                               -
zroot  utf8only              on                              -
zroot  normalization         formD                           -
zroot  casesensitivity       sensitive                       -
zroot  vscan                 off                             default
zroot  nbmand                off                             default
zroot  sharesmb              off                             default
zroot  refquota              none                            default
zroot  refreservation        none                            default
zroot  guid                  15284767340242609850            -
zroot  primarycache          all                             default
zroot  secondarycache        all                             default
zroot  usedbysnapshots       0B                              -
zroot  usedbydataset         98K                             -
zroot  usedbychildren        366G                            -
zroot  usedbyrefreservation  0B                              -
zroot  logbias               latency                         default
zroot  objsetid              54                              -
zroot  dedup                 off                             default
zroot  mlslabel              none                            default
zroot  sync                  standard                        default
zroot  dnodesize             auto                            local
zroot  refcompressratio      1.00x                           -
zroot  written               98K                             -
zroot  logicalused           415G                            -
zroot  logicalreferenced     49K                             -
zroot  volmode               default                         default
zroot  filesystem_limit      none                            default
zroot  snapshot_limit        none                            default
zroot  filesystem_count      none                            default
zroot  snapshot_count        none                            default
zroot  snapdev               hidden                          default
zroot  acltype               posix                           local
zroot  context               none                            default
zroot  fscontext             none                            default
zroot  defcontext            none                            default
zroot  rootcontext           none                            default
zroot  relatime              off                             local
zroot  redundant_metadata    all                             default
zroot  overlay               on                              default
zroot  encryption            aes-256-gcm                     -
zroot  keylocation           file:///etc/zfs/keys/zroot.key  local
zroot  keyformat             passphrase                      -
zroot  pbkdf2iters           350000                          -
zroot  encryptionroot        zroot                           -
zroot  keystatus             available                       -
zroot  special_small_blocks  0                               default
```

```
$ zpool list -v zroot
NAME        SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
zroot       448G   366G  81.5G        -         -    76%    81%  1.00x    ONLINE  -
  sda3      449G   366G  81.5G        -         -    76%  81.8%      -    ONLINE
```
